src/hotspot/share/classfile/stackMapTable.hpp

```diff
@@ -141,15 +153,40 @@ class StackMapReader : StackObj {
 
  public:
   // Constructor
-  StackMapReader(ClassVerifier* v, StackMapStream* stream, char* code_data,
-                 int32_t code_len, TRAPS);
+  StackMapReader(ClassVerifier* v, StackMapStream* stream,
+                 char* code_data, int32_t code_len,
+                 StackMapFrame* init_frame,
+                 u2 max_locals, u2 max_stack, TRAPS);
+
+  inline int32_t get_frame_count()  const { return _frame_count; }
+  inline StackMapFrame* frame()     const { return _frame_buf; }
+  inline StackMapFrame* pre_frame() const { return _pre_frame; }
+  inline char* code_data()          const { return _code_data; }
+  inline int32_t code_length()      const { return _code_length; }
+  inline bool at_end() { return _stream->at_end(); }
+
+  inline void set_pre_frame(StackMapFrame* frame) { _pre_frame = frame; }
+  StackMapFrame* next_helper(TRAPS);
+  StackMapFrame* next(TRAPS);
+
+  void check_offset(StackMapFrame* frame, TRAPS) {
+    int offset = frame->offset();
+    if (offset >= _code_length || _code_data[offset] == 0) {
+      _verifier->verify_error(
+          ErrorContext::bad_stackmap(0, frame),
+          "StackMapTable error: bad offset");
+    }
```

Comment:
Rather than have this inlined in the header file, can you put the definition right before where it's used?  I think I only found one place.  I believe this is in only one place.

next_helper should be declared private.